feat: Change Feeds Enabling/Writing/Reading

mumoshu opened the pull request:
## What is the motivation?

We want a feature that serves all kinds of "CDC" use-cases.

## What does this change do?

This pull request adds the initial version of the so-called "Change Feeds" feature which almost provides a way
to reliably read the change history of a SurrealDB table (not the instance/cluster or the namespace. Database change feeds is WIP though).

Note that this initial version uses the "sound" variant of Versionstamps for Tikv, RocksDB, InMem, IndexedDB and SpeeDB. It wraps incrementing the per-database counter within the transaction to provide a monotonic Versionstamp for those backends.

We know this causes a lot of contention on the counter backed by the underlying KVS and our plan is to have unsound Versionstamp implementations in case the user wants to have more performance than the correctness of change feeds.

The good(I suppose) news is that the FoudnationDB backend uses the `set versionstamped key` atomic operation provided natively by the FoundationDB for Versionstamps. It's lockless and mostly contention-less (in terms of we don't need to "guard" the counter using the transaction by ourselves) and supposed to be more performance than other distributed SurrealDB backends.

## What is your testing strategy?

I've written a unit and integration (API) tests that checks the behavior of Change Feeds for the following backends:

- RocksDB
- InMem
- TiKV
- FoundationDB
- SpeeDB

Honestly speaking, I'm not sure how I can test the IndexedDB in an automated way!
Suggestions on how we test it are much appreciated.

## Is this related to any issues?

N/A

Probably we'd better have a discussion/issue for any design/usage/implmenetation/etc discussions once the first implementation turned out to work.

## Have you read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)?

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

rushmorem commented:
> Honestly speaking, I'm not sure how I can test the IndexedDB in an automated way!

We could add a module for Wasm tests inside `lib/tests/api/` and use `wasm-bindgen-test` to test those.

mumoshu commented on lib/tests/changefeeds.rs:
```diff
@@ -0,0 +1,159 @@
+	let sql = "
+        DEFINE TABLE person CHANGEFEED 1h;
+		DEFINE FIELD name ON TABLE person
+			ASSERT
+				IF $input THEN
+					$input = /^[A-Z]{1}[a-z]+$/
+				ELSE
+					true
+				END
+			VALUE
+				IF $input THEN
+					'Name: ' + $input
+				ELSE
+					$value
+				END
```
I've copy-pasted this conditional value transformation thing from another test.
The good part of having this here is that we can ensure in the test that change feed captures the value after the conversion!

mumoshu commented on lib/tests/api/mod.rs:
```diff
@@ -671,6 +671,139 @@ async fn delete_record_range() {
 }
 
+#[tokio::test]
+async fn changefeed() {
```
By adding this, we ensure that the change feed work across the following backends:

- TiKV
- FDB
- SpeeDB
- RocksDB
- InMem

mumoshu commented on lib/src/vs/oracle.rs:
```diff
@@ -0,0 +1,167 @@
+//! System time based versionstamp.
```
This module is used from nowhere for now.

However, this will be used as an alternative versionstamp producer for InMem, RocksDB, IndexedDB, and perhaps SpeeDB too.

For InMem and RocksDB, we will use `SysTimeCounter` as the runtime environment of a SurrealDB instance backed by either KVS should have a monotonic system clock, which `SysTimeCounter` relies on.

For IndexedDB, we will use `EpochCounter` which does not depend on the system clock.

phughk replied:
Yes, I think I could move ds/cl.rs Timestamp into this as well potentially; We may want to call it lib/src/time instead of vs. Not sure.
I need timestamp arithmetic that is synchronised across a cluster. So oracle timestamps T1, T2 and then `T1+5s<T2` for example. I know that's not how the hybrid logical clock works. But we need something centralised that is temporal instead of logical.

mumoshu commented on lib/src/sql/statement.rs:
```diff
@@ -247,4 +252,22 @@ mod tests {
+
+	#[test]
+	fn show_database_changes() {
+		let sql = "SHOW CHANGES FOR DATABASE SINCE 123456";
```
Change Feed in this pull request does support reading the feed for the all the tables in a database via this query.

However, the write side does not support the database-wide change feed. That is, there's no facility to let you enable the database-wide change feed recording.

Maybe we need `DEFINE DATABSE CHANGEFEED 1h` or something like that implemented in an upcoming PR to enable the write side of the database change feed!

mumoshu commented on lib/src/kvs/mem/mod.rs:
```diff
@@ -101,6 +114,75 @@ impl Transaction {
 		// Return result
 		Ok(res)
 	}
+	/// Obtain a new change timestamp for a key
```
`get_timestamp` and `set_versionstamped_key` still have a lot of code duplication across RocksDB, IndexedDB, InMem, and SpeeDB now.

However, as we already have full test coverage for each implementation, I think it's good to go for now and do a refactoring later(if necessary).

WDYT?

phughk replied:
Definitely good to go. This change is already quite big and we want to iterate, not land big changes IMO. We can refactor in the future. I have timestamp things I want to reconcile with this change as well :)

mumoshu commented on lib/src/dbs/executor.rs:
```diff
@@ -54,11 +54,17 @@ impl<'a> Executor<'a> {
 		}
 	}
 
+	/// Commits theh transaction if it is local.
+	/// This function takes two additional parameters that are not present in the underlyingn kvs::Transaction::commit() function:
+	/// ns: The namespace of the transaction
+	/// db: The database of the transaction
+	/// These parameters are used updating the changefeed.
+	///
```
I believe this is the "biggest" change to the existing codebase in this PR although the LOC change isn't actually large.

Presuming we do a commit per transaction that is tied to a namespace and a database, we take those two as the parameters, so that we can transactional update the change feed(which is, by its nature, tied to a namespace and a database)

mumoshu replied:
@phughk It turned out we (still) support cross namepace and cross database transactions 🤦
Honestly speaking, I don't know if it's practical in the long term.
Anyway, I reverted the changes to the commit function's signature and instead modified the change feed recorder to buffer changes along with the namespace and the database used when the change is made.
That way, although you don't have a single change feeds that spans namespaces and databases, you still have multiple database change feeds that can be concurrently updated via a single transaction.

mumoshu commented on lib/Cargo.toml:
```diff
@@ -106,8 +106,8 @@ sha2 = "0.10.6"
 thiserror = "1.0.40"
-tikv = { version = "0.1.0", package = "tikv-client", optional = true }
-tikv-client-proto = { version = "0.1.0", optional = true }
+tikv = { version = "0.1.0-surreal.1", package = "surrealdb-tikv-client", optional = true }
+tikv-client-proto = { version = "0.1.0-surreal.1", package = "surrealdb-tikv-client-proto", optional = true }
```
I've updated our fork of TiKV rust client to have some additional functionalities needed by Change Feed:

https://github.com/surrealdb/tikv-client/tree/0.1.0-surreal.x

and published all the related crates to crates.io so that we can depend on the crates this way

https://crates.io/search?q=surrealdb-tikv

I didn't use `git` dependency in Cargo.toml as that would prevent us from publishing surrealdb as a crate.

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories

mumoshu commented on lib/src/cf/mod.rs:
```diff
@@ -0,0 +1,7 @@
+pub(crate) mod writer;
+
+pub use self::gc::*;
+pub use self::reader::read;
```
`SHOW CHANGES` query uses this `read` function under the hood.

mumoshu commented on lib/src/cf/mod.rs:
```diff
@@ -0,0 +1,7 @@
+
+pub use self::gc::*;
+pub use self::reader::read;
+pub use self::writer::Writer;
```
This is what we use for recording changes.

More concretely, our `Transaction` now holds a reference to a single instance of this `Writer` and each create/update/delete to a document results in a function call to the Writer instance for "buffering changes" until the commit.

arriqaaq commented on lib/src/cf/gc.rs:
```diff
@@ -0,0 +1,53 @@
+	limit: Option<u32>,
+) -> Result<(), Error> {
+	let beg = cf::ts_prefix(ns, db, vs::u64_to_versionstamp(0));
+	let end = cf::ts_prefix(ns, db, vs::u64_to_versionstamp(watermark + 1));
```
Just curious, based on the comment for `gc_db`, it seems to delete all the entries < watermark right? But it says older than given watermark?

mumoshu replied:
Awesome! I got messed up while implementing this... Fixed the implementation and the test as that was my intention!
5dee201eb365659472bad5f42acb47a711e106a2

arriqaaq replied:
Thank you!

arriqaaq commented on lib/src/cf/reader.rs:
```diff
@@ -0,0 +1,93 @@
+	visible
+}
+
+// put_change writes all the mutations buffered for this transaction
```
This comment is probably for another function?

arriqaaq commented on lib/src/vs/oracle.rs:
```diff
@@ -0,0 +1,167 @@
+
+impl EpochCounter {
+	pub fn now(&mut self) -> Versionstamp {
+		let counter = self.counter.fetch_add(1, Ordering::SeqCst);
```
Rust noob here, I saw that fetch_add (Adds to the current value, returning the previous value). This means it is returning the previous value of counter, and not the current value. Am i right in assuming this? Could lead to stale time?

mumoshu replied:
@arriqaaq Thanks! You are absolutely correct on how fetch_add works.

I think this implementation is correct though, because `now()` is the only way to grab a new versionstamp. By using fetch_add this way, we basically treat the "previous value" as the current versionstamp and return it, preparing the next versionstamp at the same time.

arriqaaq replied:
Oh I see, thank you for the explanation!

arriqaaq replied:
Can we just add a unit test case just to verify this?

mumoshu replied:
Hey @arriqaaq, Thanks a lot for your confirmation! Re: test, does this one seem reasonable to you?
https://github.com/surrealdb/surrealdb/blob/b775bdab05d3a0f80df954a103a321dd4df06b7d/lib/src/vs/oracle.rs#L153-L169

arriqaaq replied:
Thank you @mumoshu, this does look reasonable, appreciate it much.

mumoshu replied:
Thanks a lot for your confirmation @arriqaaq!

arriqaaq commented on lib/src/vs/oracle.rs:
```diff
@@ -0,0 +1,167 @@
+			self.stale = (current_physical_time, current_logical_time);
+			u64_u16_to_versionstamp(current_physical_time, current_logical_time)
+		} else {
+			u64_u16_to_versionstamp(self.stale.0, self.stale.1)
```
Does this mean in case when a mutex lock fails, we return back older version of timestamp? Wouldn't that be inconsistent as it would be returning an already existing timestamp which could have been used?

mumoshu replied:
@arriqaaq Thanks! You are correct! And, yes, they can be inconsistent.

But do note that even though this oracle provided no stale verionstamps, you still need to separately ensure that the transactions are committed with consistent order and unique motonic versionstamps.

I.e. you need to separately ensure that the commits are made in the order of unique versionstamps and I presume that's not strictly the responsibility of this package.

That said, I presume we have two other options; Do you prefer one of those or have another idea?

- Return `Err(PleaseRetryTheFunCall)` on L12 so that the caller of `now()` can retry calling it until succeed (with or without aborting the whole transaction, depends on how you impl it). The resulting versionstamps are guaranteed to be monotonic and unique.
- Make this `async` and wait until the lock succeeds. The resulting versionstamps are guaranteed to be monotonic and unique.

arriqaaq replied:
Thank you for the explanation, I understand the challenges here. I think both approaches are nice, I strongly don't have a preference because as you've said, it is better to pass on the responsibility to the caller on how to handle it.

mumoshu commented:
Fixed all new clippy errors and updated all `clone_transction()` calls to `try_clone_transaction()` following recent renaming!
This should be good to go now.

sgirones commented on lib/src/dbs/executor.rs:
```diff
@@ -56,11 +56,17 @@ impl<'a> Executor<'a> {
 		}
 	}
 
+	/// Commits theh transaction if it is local.
```
```suggestion
	/// Commits the transaction if it is local.
```

sgirones commented on lib/src/dbs/executor.rs:
```diff
@@ -56,11 +56,17 @@ impl<'a> Executor<'a> {
 	}
 
+	/// Commits theh transaction if it is local.
+	/// This function takes two additional parameters that are not present in the underlyingn kvs::Transaction::commit() function:
```
```suggestion
	/// This function takes two additional parameters that are not present in the underlying kvs::Transaction::commit() function:
```

sgirones commented on lib/src/kvs/speedb/mod.rs:
```diff
@@ -136,6 +137,60 @@ impl Transaction {
+
+		let verbytes = u64_to_versionstamp(ver);
+
+		self.tx.lock().await.as_ref().unwrap().put(k, verbytes)?;
```
do you need to check for `self.tx` to be writable? if yes, I guess it applies to all the `get_timestamp` methods defined for other datastores

mumoshu replied:
Great question! Yeah probably. I honestly don't fully understand the consequence of having these prechecks in the beginning of functions that writes KVs! To make things consistent with existing functions though, adding the writeability check does make sense.

In the latest commit, I fixed `get_versionstamped_key` rather than `get_timestamp`. The former is supposed to be used only in a writeable transaction whereas the latter can be used even in a read-only transaction.

sgirones commented on lib/src/kvs/tx.rs:
```diff
@@ -394,6 +399,120 @@ impl Transaction {
+		K: Into<Key>,
+	{
+		#[cfg(debug_assertions)]
+		trace!(target: LOG, "Get Timestamp");
```
is it worth adding the Key in the trace message?

mumoshu replied:
Yeah I believe so! Added. Thanks for the suggestion ☺️

sgirones commented on lib/src/kvs/tx.rs:
```diff
@@ -1909,4 +2030,47 @@ impl Transaction {
+	// complete_changes will complete the changefeed recording for the given namespace and database.
+	//
+	// Under the hood, this function calls the transaction's `set_versionstamped_key` for each change.
+	// Every change must be recorded by calling this struct's `change` function beforehand.
```
```suggestion
	// Every change must be recorded by calling this struct's `record_change` function beforehand.
```

sgirones commented on lib/src/kvs/tx.rs:
```diff
@@ -1909,4 +2030,47 @@ impl Transaction {
+	//
+	// Under the hood, this function calls the transaction's `set_versionstamped_key` for each change.
+	// Every change must be recorded by calling this struct's `change` function beforehand.
+	// If there was no preceeding `change` function calls for this transaction, this function will do nothing.
```
```suggestion
	// If there was no preceeding `record_change` function calls for this transaction, this function will do nothing.
```

sgirones commented on lib/src/kvs/tx.rs:
```diff
@@ -1909,4 +2030,47 @@ impl Transaction {
+	// This function should be called immediately before calling the commit function to guarantee that
+	// the lock, if needed by lock=true, is held only for the duration of the commit, not the entire transaction.
+	//
+	// This function is here becauses it needs access to mutably borrow the transaction.
```
```suggestion
	// This function is here because it needs access to mutably borrow the transaction.
```

sgirones commented on lib/src/sql/statements/show.rs:
```diff
@@ -0,0 +1,206 @@
+		// Allowed to run?
+		opt.check(Level::Db)?;
+		// Allowed to run?
+		opt.check(Level::No)?;
```
this second check is not really useful, because if line 34 succeed, this will always succeed.

mumoshu replied:
Thanks for pointing it out! Removed L35-36!

mumoshu commented:
@sgirones Thank you for reviewing! I've addressed everything in b775bdab05d3a0f80df954a103a321dd4df06b7d ☺️

mumoshu commented:
I believe I've addressed all the comments! This should be good to go now.

phughk commented on lib/src/kvs/mem/mod.rs:
```diff
@@ -101,6 +114,84 @@ impl Transaction {
+		k.append(&mut ts.to_vec());
+		k.append(&mut suffix.clone());
+
+		println!(
```
We should use trace! wherever possible IMO. We can also add parameters to the trace command so they can be filtered on capture.

mumoshu replied:
Thanks again! Modified to use `trace!` in https://github.com/surrealdb/surrealdb/pull/2306

phughk reviewed (approved):
LGTM! Thanks Yusuke! Definitely loads to read, but the approach is sensible

mumoshu commented on lib/src/cf/reader.rs:
```diff
@@ -0,0 +1,93 @@
+		Some(x) => cf::ts_prefix(ns, db, vs::u64_to_versionstamp(x)),
+		None => {
+			let ts = tx.get_timestamp(seq, false).await?;
+			cf::ts_prefix(ns, db, ts)
```
Thanks @phughk! Changed to `prefix_ts` in https://github.com/surrealdb/surrealdb/pull/2306

mumoshu commented on lib/src/cf/reader.rs:
```diff
@@ -0,0 +1,93 @@
+	let mut r = Vec::<ChangeSet>::new();
+	// iterate over _x and put decoded elements to r
+	for (k, v) in _x {
+		println!("read: k={}", show(k.as_slice()));
```
Thanks! Updated to use `trace!` in https://github.com/surrealdb/surrealdb/pull/2306 ☺️

mumoshu commented on lib/src/kvs/speedb/mod.rs:
```diff
@@ -136,6 +137,60 @@ impl Transaction {
+				let prev = to_u64_be(array);
+				prev + 1
+			}
+			None => 1,
```
Correct! It's written back in the subsequent `put` within the same function.

mumoshu commented on lib/src/cf/writer.rs:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+// fn tb_mutations(ts: u64, tb: &str, changes: Vec<(ID, Patch))
```
Wow good catch! Removed in https://github.com/surrealdb/surrealdb/pull/2306

tobiemh commented:
Superseded by #2306 🎉
